### PR TITLE
refactors climbing into an element

### PR DESCRIPTION
**Pull request by ghost**
:cl:
refactor: climbing is now done through an element
/:cl:

this moves climbing behavior from structures to an element, this PR was made for 3/4th walls because i got sad when i saw the code

### CODE REVIEW

**Comment by Qustinnus** [comment]
@actioninja i can keep my testicles

File: code/game/objects/structures.dm [hunk 1]
```diff
@@ -8,13 +8,7 @@
...
 	receive_ricochet_chance_mod = 0.6
 	pass_flags_self = PASSSTRUCTURE
-
-	var/climb_time = 20
-	var/climb_stun = 20
-	var/climbable = FALSE
-	var/mob/living/structureclimber
-	var/broken = 0 //similar to machinery's stat BROKEN
-
+	var/broken
```
Comment (Fikou) [inline code/game/objects/structures.dm:11]:
```suggestion
	var/broken = FALSE
```

File: code/game/objects/structures/crates_lockers/closets.dm [hunk 2]
```diff
@@ -166,7 +166,6 @@
 	opened = TRUE
 	if(!dense_when_open)
 		density = FALSE
-	climb_time *= 0.5 //it's faster to climb onto an open thing
```
Comment (Fikou) [inline code/game/objects/structures/crates_lockers/closets.dm:169]: lame removal

File: code/datums/elements/climbable.dm [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+/datum/element/climbable/proc/mousedrop_receive(atom/climbed_thing, atom/movable/dropped_atom, mob/user)
+	SIGNAL_HANDLER
+	if(user == dropped_atom && isliving(dropped_atom))
+		var/mob/living/living_target = dropped_atom
+		if(isanimal(living_target))
+			var/mob/living/simple_animal/animal = dropped_atom
+			if (!animal.dextrous)
+				return
+		if(living_target.mobility_flags & MOBILITY_MOVE)
+			climb_structure(climbed_thing, living_target)
```
Comment (Ghommie) [inline code/datums/elements/climbable.dm:94]:
```suggestion
			INVOKE_ASYNC(src, .proc/climb_structure, climbed_thing, living_target)
```

Comment (Qustinnus) [reply code/datums/elements/climbable.dm:94]: cheers saves me from doing it <3

File: code/game/objects/structures/crates_lockers/closets.dm (hunk 2 above)
Comment (Qustinnus) [reply code/game/objects/structures/crates_lockers/closets.dm:169]: yeah i wasnt sure how to fix this but i thought of a way

**Comment by Qustinnus** [comment]
@Cyberboss why are you turning this finished PR into a draft when it compiles

**Comment by actioninja** [comment]
conflict

**Comment by Qustinnus** [comment]
third time i fixed them

File: code/datums/elements/climbable.dm [hunk 4]
```diff
@@ -0,0 +1,104 @@
+/datum/element/climbable
+	element_flags = ELEMENT_BESPOKE|ELEMENT_DETACH
+	id_arg_index = 2
+	///Time it takes to climb onto the object
+	var/climb_time = 20
+	///Stun duration for when you get onto the object
+	var/climb_stun = 20
```
Comment (Jared-Fogle) [inline code/datums/elements/climbable.dm:7]:
```suggestion
	var/climb_time = (2 SECONDS)
	///Stun duration for when you get onto the object
	var/climb_stun = (2 SECONDS)
```

File: code/datums/elements/climbable.dm [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+	UnregisterSignal(target, list(COMSIG_ATOM_ATTACK_HAND, COMSIG_PARENT_EXAMINE, COMSIG_MOUSEDROPPED_ONTO, COMSIG_ATOM_BUMPED))
+	return ..()
+
+/datum/element/climbable/proc/on_examine(atom/source, mob/user, list/examine_texts)
+	SIGNAL_HANDLER
+
+	if(can_climb(source, user))
+		examine_texts += "<span class='notice'>[source] looks climbable</span>"
+
+/datum/element/climbable/proc/can_climb(atom/source, mob/user)
```
Comment (Jared-Fogle) [inline code/datums/elements/climbable.dm:36]: Is this overridden at any point? Or is this just for future proofing?

Comment (Qustinnus) [reply code/datums/elements/climbable.dm:36]: its just encapsulation really, I dislike jamming everything into one proc

**Pull request merged**